pageserver/src/restore_local_repo.rs

```diff
@@ -0,0 +1,460 @@
+//
+// Restore chunks from local Zenith repository
+//
+// This runs once at Page Server startup. It loads all the "snapshots" and all
+// WAL from all timelines from the local zenith repository into the in-memory page
+// cache.
+//
+// This also initializes the "last valid LSN" in the page cache to the last LSN
+// seen in the WAL, so that when the WAL receiver is started, it starts
+// streaming from that LSN.
+//
+
+use log::*;
+use regex::Regex;
+use std::fmt;
+
+use std::error::Error;
+use std::fs;
+use std::fs::File;
+use std::io::Read;
+use std::io::Seek;
+use std::io::SeekFrom;
+use std::path::{Path, PathBuf};
+use std::cmp::max;
+
+use anyhow::Result;
+use bytes::Bytes;
+
+use crate::page_cache;
+use crate::page_cache::PageCache;
+use crate:: PageServerConf;
+use crate::page_cache::BufferTag;
+use crate::waldecoder::WalStreamDecoder;
+use crate::ZTimelineId;
+
+
+// From pg_tablespace_d.h
+//
+// FIXME: we'll probably need these elsewhere too, move to some common location
+const DEFAULTTABLESPACE_OID: u32 = 1663;
+const GLOBALTABLESPACE_OID: u32 = 1664;
+
+//
+// Load it all into the page cache.
+//
+pub fn restore_timeline(conf: &PageServerConf, pcache: &PageCache, timeline: ZTimelineId) -> Result<()> {
+
+    let timelinepath = PathBuf::from("timelines").join(&timeline.to_str());
+
+    if !timelinepath.exists() {
+        anyhow::bail!("timeline {} does not exist in the page server's repository");
+    }
+
+    // Scan .zenith/timelines/<timeline>/snapshots
+    let snapshotspath = "timelines/".to_owned() + &timeline.to_str() + "/snapshots";
+
+    let mut last_snapshot_lsn: u64 = 0;
+
+    for direntry in fs::read_dir(&snapshotspath).unwrap() {
+        let filename = direntry.unwrap().file_name().to_str().unwrap().to_owned();
+
+        let lsn = u64::from_str_radix(&filename, 16)?;
+        last_snapshot_lsn = max(lsn, last_snapshot_lsn);
+
+        restore_snapshot(conf, pcache, timeline, &filename)?;
+        info!("restored snapshot at {}", filename);
+    }
+
+    if last_snapshot_lsn == 0 {
+        error!("could not find valid snapshot in {}", &snapshotspath);
+        // TODO return error?
+    }
+    pcache.init_valid_lsn(last_snapshot_lsn);
+
+    restore_wal(conf, pcache, timeline, last_snapshot_lsn)?;
+
+    Ok(())
+}
+
+pub fn find_latest_snapshot(_conf: &PageServerConf, timeline: ZTimelineId) -> Result<u64> {
+
+    let snapshotspath = format!("timelines/{}/snapshots", timeline);
```

Comment:
Is this supposed to be a relative path?

The by-the-book way to build this path would probably be
```
let snapshotspath: PathBuf = ["timelines", &timeline.to_string(), "snapshots"].iter().collect();
```
which I find a bit clunky. This would also work:
```
 let snapshotspath = PathBuf::from("timelines").join(timeline.to_string()).join("snapshots");
```